src/core/filesystem.cc

```diff
@@ -224,10 +228,54 @@ LocalFileSystem::DownloadFileFolder(
   return Status::Success;
 }
 
+int
+LocalFileSystem::IsPathDirectory(const char* path)
+{
+  struct stat s_buf;
+
+  if (stat(path, &s_buf))
+    return 0;
+
+  return S_ISDIR(s_buf.st_mode);
+}
+
+int
+LocalFileSystem::DeleteFolderRecursive(const std::string& path)
+{
+  struct dirent* ep;
+  DIR* dp = opendir(path.c_str());
+
+  while ((ep = readdir(dp)) != NULL) {
+    if (strcmp(ep->d_name, ".") == 0 || strcmp(ep->d_name, "..") == 0) {
+      continue;
+    }
+    std::string tmp_path = path + "/" + std::string(ep->d_name);
+    if (IsPathDirectory(tmp_path.c_str())) {
+      DeleteFolderRecursive(tmp_path);
+    } else {
+      if (remove(tmp_path.c_str()) != 0) {
+        return -1;
+      }
+    }
+  }
+
+  closedir(dp);
+  return rmdir(path.c_str());
+}
+
 Status
 LocalFileSystem::DestroyFileFolder(const std::string& path)
 {
-  // Do nothing
+  if (IsPathDirectory(path.c_str())) {
```

Comment:
You are deleting the local copy of the model?!? Isn't this going to delete the model repos? Including our repos on dldata? This method should be named something like ReleaseDownloadFileFolder to make it clear that it is not (always) deleting/destroying.

Reply:
This is only deleted in the case of the path matching the format of `"/tmp/file*"` i.e. in the case the file is a local copy of a file in a cloud repository.

Reply:
Fixed naming to ReleaseDownloadFileFolder

Reply:
Why make a copy of a local filesystem model? Why is the custom backend any different than other cases?

Reply:
The behavior for Custom Backend is the same as what it was for other backends and has not changed. What has changed (fixed) is the cleanup of temporary files that are created when downloading from S3. 

The ReleaseDownloadFileFolder of the local filesystem clears models that were created in by calling DownloadFileFolder in S3. It is confusing as first but as @Guan mentioned the file being released resides in the local filesystem and not S3.

Reply:
I think it was confusing because of the naming of the functions. Even DownloadFileFolder should probably change to LocalizeFileFolder or similar. You can have LocalizeFileFolder return a shared_ptr object which represents the temp local copy and then have the destructor of that object remove the directory... then you wouldn't need ReleaseDownloadFileFolder and the ownership and functionality would be clearer.

Reply:
The new class/object can be called TemporaryDirectory or similar. This will also fix some bugs we have now where if errors occur the tmpdir is not cleaned up because DestroyFileFolder is not called.

Reply:
Sure. I will make these changes